Riot/Modules/Onboarding/AuthenticationCoordinator.swift

```diff
@@ -0,0 +1,418 @@
+        
+    }
+    
+    // MARK: - Registration Handlers
+    /// Determines the next screen to show from the flow result and pushes it.
+    func showNextScreen(for flowResult: FlowResult) {
+        // TODO
+    }
+    
+    /// Handles the creation of a new session following on from a successful authentication.
+    func onSessionCreated(session: MXSession, isAccountCreated: Bool) {
+        registerSessionStateChangeNotification(for: session)
+        
+        self.session = session
+        // self.password = password
+        
+        if canPresentAdditionalScreens {
+            showLoadingAnimation()
+        }
+        
+        completion?(.didLogin(session: session, authenticationType: isAccountCreated ? .register : .login))
+    }
+    
+    // MARK: - Additional Screens
+    
+    /// Replace the contents of the navigation router with a loading animation.
+    private func showLoadingAnimation() {
+        let loadingViewController = LaunchLoadingViewController()
+        loadingViewController.modalPresentationStyle = .fullScreen
+        
+        // Replace the navigation stack with the loading animation
+        // as there is nothing to navigate back to.
+        navigationRouter.setRootModule(loadingViewController)
+    }
+    
+    /// Present the key verification screen modally.
+    private func presentCompleteSecurity() {
+        guard let session = session else {
+            MXLog.error("[LegacyAuthenticationCoordinator] presentCompleteSecurity: Unable to present security due to missing session.")
+            authenticationDidComplete()
+            return
+        }
+        
+        let isNewSignIn = true
+        let cancellable = !session.vc_homeserverConfiguration().encryption.isSecureBackupRequired
+        let keyVerificationCoordinator = KeyVerificationCoordinator(session: session, flow: .completeSecurity(isNewSignIn), cancellable: cancellable)
+        
+        keyVerificationCoordinator.delegate = self
+        let presentable = keyVerificationCoordinator.toPresentable()
+        presentable.presentationController?.delegate = self
+        navigationRouter.present(presentable, animated: true)
+        keyVerificationCoordinator.start()
+        add(childCoordinator: keyVerificationCoordinator)
+    }
+    
+    /// Complete the authentication flow.
+    private func authenticationDidComplete() {
+        completion?(.didComplete)
+    }
+    
+    private func registerSessionStateChangeNotification(for session: MXSession) {
+        NotificationCenter.default.addObserver(self, selector: #selector(sessionStateDidChange), name: .mxSessionStateDidChange, object: session)
+    }
+
+    private func unregisterSessionStateChangeNotification() {
+        NotificationCenter.default.removeObserver(self, name: .mxSessionStateDidChange, object: nil)
+    }
+                                      
+    @objc private func sessionStateDidChange(_ notification: Notification) {
```

Comment:
Code in this `sessionStateDidChange` seems pretty important, but also quite hard to read at the moment, which I imagine means hard to test as well. Additionally I am not sure it is the direct responsibility of the Coordinator. Could we extract this into some nested object that would focus on this state setup, thus making it easier to test?

Reply:
I've created a `SessionVerificationListener` class (open to a better suggestion on the name) and moved this code in here instead. It was also a duplication of the code in `LegacyAuthenticationCoordinator` so that has fixed that issue too 🙏

In terms of readability, this code was a 1:1 conversion of existing [Objective-C](https://github.com/vector-im/element-ios/blob/94dab7ad4b19f67aaea324cdceda87f2ed5c686b/Riot/Modules/Authentication/AuthenticationViewController.m#L1394) that I wasn't too confident about altering the structure as there looked to be a lot of complexity hidden away in there that could potentially be broken.

Reply:
Yea keeping the 1:1 is a good idea, it just looked really messy inside the coordinator. Taking the code out really helped 👍